src/adapters/webpush/mod.rs

```diff
@@ -0,0 +1,359 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+extern crate serde_json;
+
+mod crypto;
+mod db;
+
+use controller::Controller;
+use foxbox_adapters::adapter::*;
+use foxbox_taxonomy::api::{ Error, InternalError };
+use foxbox_taxonomy::values::{ Range, Type, Value, Json };
+use foxbox_taxonomy::services::*;
+use foxbox_taxonomy::util::Id;
+use hyper::header::{ ContentEncoding, Encoding };
+use hyper::Client;
+use hyper::client::Body;
+use rusqlite::{ self };
+use std::collections::{ HashMap, HashSet };
+use std::sync::Arc;
+use std::thread;
+use transformable_channels::mpsc::*;
+
+header! { (Encryption, "Encryption") => [String] }
+header! { (EncryptionKey, "Encryption-Key") => [String] }
+
+static ADAPTER_NAME: &'static str = "WebPush adapter (built-in)";
+static ADAPTER_VENDOR: &'static str = "team@link.mozilla.org";
+static ADAPTER_VERSION: [u32;4] = [0, 0, 0, 0];
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub struct Subscription {
+    pub push_uri: String,
+    pub public_key: String
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub struct SubscriptionGetter {
+    subscriptions: Vec<Subscription>
+}
+
+impl SubscriptionGetter {
+    fn new(subs: Vec<Subscription>) -> Self {
+        SubscriptionGetter {
+            subscriptions: subs
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub struct ResourceGetter {
+    resources: Vec<String>
+}
+
+impl ResourceGetter {
+    fn new(res: Vec<String>) -> Self {
+        ResourceGetter {
+            resources: res
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub struct NotifySetter {
+    resource: String,
+    message: String
+}
+
+impl Subscription {
+    fn notify(&self, message: &String) {
+        let enc = match self::crypto::encrypt(&self.public_key, message.clone()) {
+            Some(x) => x,
+            None => {
+                warn!("notity subscription {} failed for {}", self.push_uri, message);
+                return;
+            }
+        };
+
+        let client = Client::new();
+        let res = match client.post(&self.push_uri)
```

Comment:
Hm. I wonder what is best to do here. I spawn a thread which sequentially executes all of the push notifications. My thinking was if we have a lot of subscriptions active, we would want to avoid hammering the push service by spawning N threads all at once. On the other hand, if one user uses another push service and that one hangs, it would block all of the others from completing...